lib/rjgit.rb

```diff
@@ -209,15 +216,32 @@ def self.grep(repository, query, options={})
 
       query = Regexp.new(query.source, query.options | Regexp::IGNORECASE) if case_insensitive
 
-      ls_tree(repo, nil, options.fetch(:ref, 'HEAD'), ls_tree_options).each_with_object({}) do |item, result|
-        blob = Blob.new(repo, item[:mode], item[:path], walk.lookup_blob(ObjectId.from_string(item[:id])))
-        next if blob.binary?
+      # We optimize below by first grepping the entire file, and then, if a match is found, then identifying the individual line.
+      # To avoid missing full-line matches during the optimization, we first convert multiline anchors to single-line anchors.
+      query = Regexp.new(query.source.gsub(/\A\\A/, '^').gsub(/\\z\z/, '$'), query.options)
```

Comment:
I'm not quite sure why this step would be necessary; can you explain why we would miss full-line matches without this step?

Reply:
From the user's perspective, we are searching each line individually for a match, the same way `git grep` works.  So a user would reasonably expect that `\A` and `\z` should refer to the beginning and end of an individual line in the file, not to the beginning and end of the entire file.

However, because we optimize below by first searching the entire file (`next unless query.match(file_contents)`), `\A` and `\z` mean the beginning and end of the entire file during that match.

For example, if a user greps for `\A123`, they mean "Any line that begins with 123".  If the first line of the file begins with `456` and the second line begins with `123`, `next unless query.match(file_contents)` will omit that file from results.  By converting to `^123`, we ensure that the 2nd line of the file matches both when we search the entire file and when we later search line-by-line.

Reply:
Thanks for the explanation!